Update codes, settings, and resources for new feature of Spell Check

WhiteHsu opened the pull request:
Update codes, settings, and resources for the issue #16 . 

You can find more detailed information in the webpage of #16 :
https://github.com/sebbrudzinski/Open-LaTeX-Studio/issues/16

Please help suggest and feedback. Thank you very much.

sebbrudzinski commented:
Thanks for checking in, I'll try to review asap

sebbrudzinski commented on OpenLaTeXStudio/Editor/pom.xml:
```diff
@@ -50,6 +50,7 @@
         <dependency>
             <groupId>org.netbeans.api</groupId>
             <artifactId>org-openide-awt</artifactId>
+            <version>${netbeans.version}</version>
```
The versions should be declared in the parent pom.xml in the dependency management section

sebbrudzinski commented on OpenLaTeXStudio/Editor/pom.xml:
```diff
@@ -98,6 +99,11 @@
         <dependency>
             <groupId>org.netbeans.api</groupId>
             <artifactId>org-openide-util</artifactId>
+            <version>${netbeans.version}</version>
```
The versions should be declared in the parent pom.xml in the dependency management section

sebbrudzinski commented on OpenLaTeXStudio/Editor/src/main/java/latexstudio/editor/toolbar/SpellCheck.java:
```diff
@@ -0,0 +1,49 @@
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        TopComponent tc = WindowManager.getDefault().findTopComponent("EditorTopComponent");
+        EditorTopComponent etc = (EditorTopComponent) tc;        
```
We have got our own factory to retrieve the top components - let's use it. The class name is TopComponentFactory and it's used in most of our classes.

sebbrudzinski commented on OpenLaTeXStudio/Editor/src/main/java/latexstudio/editor/EditorTopComponent.java:
```diff
@@ -276,6 +287,51 @@ public void commentOutText() {
+                    if (rule instanceof SpellingCheckRule) {
+                        ((SpellingCheckRule)rule).acceptPhrases(Arrays.asList("tex", "Tex", "TEX", "documentclass", "maketitle"));  //Accept some TEX terms
+                    }
+                }
```
The LanguageTool object and all the rules should not be getting created/configured each time I click the "check" button. Ideally this should live in a separate class and be initiated at the application startup. During runtime, we would only call "check" on the tool. I think this is the main reason the app hangs for some time after clicking "check"

sebbrudzinski commented on OpenLaTeXStudio/Editor/src/main/java/latexstudio/editor/EditorTopComponent.java:
```diff
@@ -276,6 +287,51 @@ public void commentOutText() {
+                JLanguageTool langTool = new JLanguageTool(new AmericanEnglish());  //Default language is: American English
+                for (Rule rule : langTool.getAllActiveRules()) {
+                    if (rule instanceof SpellingCheckRule) {
+                        ((SpellingCheckRule)rule).acceptPhrases(Arrays.asList("tex", "Tex", "TEX", "documentclass", "maketitle"));  //Accept some TEX terms
```
We've got they key LaTeX terms in a resource file tex.cwl that is used for auto-complete. We could also use it to add exception phrases here, but this is optional for now.

sebbrudzinski commented on OpenLaTeXStudio/Editor/src/main/java/latexstudio/editor/EditorTopComponent.java:
```diff
@@ -276,6 +287,51 @@ public void commentOutText() {
+                        }                        
+                } else {  //If highlight is already enabled before clicking
+                    Highlighter.Highlight[] highlighters = highlighter.getHighlights();                
+                    for (int i = 0; i < highlighters.length; i++) {
```
There's a method called "removeAllHighlights". Perhaps it's enough to call it, instead of manually iterating through all highlights?

sebbrudzinski commented:
So, there's definitely a huge room for improvement here, but this can be tackled in new tickets. Once this is merged we can create tickets for having spell check in real time, changing the language (only English works now), changing the highlight color (low priority) etc.

I've left priority comments that must be tackled before merging inline.

WhiteHsu commented:
@sebbrudzinski 
Thanks for your comments. I totally agree that this feature can be improved largely. Because this's my first time to implement spelling check feature, I've just implemented the fundamental feature in this ticket. Once finishing this ticket, we can create more others for further enhancements.

I've also updated codes based on your review comments. Please kindly help check again and provide your feedback. Thank you very much.  :-)

sebbrudzinski commented on OpenLaTeXStudio/Editor/src/main/java/latexstudio/editor/toolbar/SpellCheck.java:
```diff
@@ -0,0 +1,49 @@
+            Exceptions.printStackTrace(ex);
+        }      
+        
+        etc.getEditorState().setDirty(true);        
```
Do we need to set the state to dirty here? The spell check itself shouldn't change the document content.

WhiteHsu replied:
@sebbrudzinski
Thanks for feedback. I've updated codes accordingly. If it's OK, please review and accept my change. Then I'll create new items and work on more enhancements for this Spell Check feature. Thank you. :-)

sebbrudzinski commented:
@WhiteHsu I've left one more minor question/comment. Otherwise this looks good enough for now.

WhiteHsu commented:
@sebbrudzinski 
Thanks for feedback. I've updated codes accordingly. If it's OK, please review and accept my change. Then I'll create new items and work on more enhancements for this Spell Check feature. Thank you.  :-)
